tests/test_overrides_parser.py

```diff
@@ -561,15 +593,41 @@ def test_key(value: str, expected: Any) -> None:
         param("false", False, id="primitive:bool"),
         # quoted string
         param(
-            "'foo \\'bar'",
+            r"'foo \'bar'",
             QuotedString(text="foo 'bar", quote=Quote.single),
             id="value:escape_single_quote",
         ),
         param(
-            '"foo \\"bar"',
+            r'"foo \"bar"',
             QuotedString(text='foo "bar', quote=Quote.double),
             id="value:escape_double_quote",
         ),
+        param(
+            r"'foo \\\'bar'",
+            QuotedString(text=r"foo \'bar", quote=Quote.single),
+            id="value:escape_single_quote_x3",
+        ),
```

Comment:
I understand that this is why you added unescaping of double backslash in _createPrimitive().
Unescaping seems like a big hammer.
Can you remind me why we can't represent `QuotedString(text=r"foo \"bar", quote=Quote.double),` with `r"'foo \\'bar'"`?

no unescaping of `\\`, only of `\'` -> `'`.

Reply:
Reading your doc change:

> which can be required e.g. for strings ending with `\`

If I understand things correctly, this looks like:
We unescaping all the double backslashes (essentially requiring double number of backslashes if downstream (OmegaConf also needs more than one backslash), and in exchange we get support for the edge case of trailing backslash (which is ambiguous with an escaped quote).

Unescaping double backslashes everywhere in a quoted string seems like a pretty big hammer.
Can we limit the effect somehow to when we really need it?

What if we had special handling for the case where a string ends with a double backslash and only unescape it in that case?
It's a bit tricky, and we will also need to apply the reverse logic when quoting a string (`with_quotes`) but I think it might be possible.

Reply:
@odelalleau  I think it should be possible:

```python
from pytest import mark


def forward(s: str) -> str:
    if s.endswith("\\\\"):
        s1 = s[0:-2].replace(r"\'", "'") + "\\"
    else:
        s1 = s.replace(r"\'", "'")
    return s1


def backward(s):
    if s.endswith("\\"):
        s1 = s.replace("'", r"\'") + "\\"
    else:
        s1 = s.replace("'", r"\'")
    return s1


@mark.parametrize(
    ("input_", "expected"),
    [
        (r"ab", r"ab"),
        (r"a\'b", r"a'b"),
        (r"ab\'", "ab'"),
        (r"\ab", r"\ab"),
        (r"ab\\ ".strip(), r"ab\ ".strip()),
        (r"ab\\'", r"ab\'"),
    ],
)
class TestEscape:
    def test_forward(self, input_, expected):
        res = forward(input_)
        assert res == expected

    def test_backward(self, input_, expected):
        res = forward(input_)
        assert res == expected
        back = backward(res)
        assert back == input_
```

Reply:
> Can you remind me why we can't represent `QuotedString(text=r"foo \"bar", quote=Quote.double),` with `r"'foo \\'bar'"`?

That's because the lexer will see a QUOTED_VALUE matching `'foo \\'` (meant to be resolved to `foo \`) and then an exception will be raised because it won't know what to do with the remaining characters `bar'` (same reason why `x = 'foo \\'bar'` in Python raises a SyntaxError).

> We unescaping all the double backslashes (essentially requiring double number of backslashes if downstream (OmegaConf also needs more than one backslash), and in exchange we get support for the edge case of trailing backslash (which is ambiguous with an escaped quote).

Essentially, yes.

> Unescaping double backslashes everywhere in a quoted string seems like a pretty big hammer.
> Can we limit the effect somehow to when we really need it?

Yes, we can limit ourselves to the `\\...\'` case (i.e., an arbitrary number of backslashes followed by a quote). This is similar to what I had in https://github.com/omry/omegaconf/pull/621 (see examples in this PR's description) -- I think if we do it in Hydra it would also make sense to do the same in OmegaConf for consistency.

> What if we had special handling for the case where a string ends with a double backslash and only unescape it in that case?

It would need to be generalized to the more general case I just mentioned but yes I think it should be doable.

Sounds good? (thanks for the sample code)

Reply:
> > Can you remind me why we can't represent `QuotedString(text=r"foo \"bar", quote=Quote.double),` with `r"'foo \\'bar'"`?
> 
> That's because the lexer will see a QUOTED_VALUE matching `'foo \\'` (meant to be resolved to `foo \`) and then an exception will be raised because it won't know what to do with the remaining characters `bar'` (same reason why `x = 'foo \\'bar'` in Python raises a SyntaxError).

So this is not because we are unescaping \\ on the resulting string (after the lexer job is done)?

> > Unescaping double backslashes everywhere in a quoted string seems like a pretty big hammer.
> > Can we limit the effect somehow to when we really need it?
> 
> Yes, we can limit ourselves to the `\\...\'` case (i.e., an arbitrary number of backslashes followed by a quote). This is similar to what I had in [omry/omegaconf#621](https://github.com/omry/omegaconf/pull/621) (see examples in this PR's description) -- I think if we do it in Hydra it would also make sense to do the same in OmegaConf for consistency.

This is not related to us using lexer modes there, right? it will just eliminate the need to use double backslash to get a backslash in quotes strings? Put up a PR and we'll see.

> > What if we had special handling for the case where a string ends with a double backslash and only unescape it in that case?
> 
> It would need to be generalized to the more general case I just mentioned but yes I think it should be doable.
> 
> Sounds good? (thanks for the sample code)

Why do you care about arbitrary number of trailing backslashes? the last one can be special (ensure we do not confuse with an escaped trailing quote), the rest can be regular and not escaped.

Reply:
> > > Can you remind me why we can't represent `QuotedString(text=r"foo \"bar", quote=Quote.double),` with `r"'foo \\'bar'"`?
> > 
> > 
> > That's because the lexer will see a QUOTED_VALUE matching `'foo \\'` (meant to be resolved to `foo \`) and then an exception will be raised because it won't know what to do with the remaining characters `bar'` (same reason why `x = 'foo \\'bar'` in Python raises a SyntaxError).
> 
> So this is not because we are unescaping \ on the resulting string (after the lexer job is done)?

No, un-escaping `foo \\'bar` would actually yield the desired `foo \'bar`.

> This is not related to us using lexer modes there, right? it will just eliminate the need to use double backslash to get a backslash in quotes strings? Put up a PR and we'll see.

Correct. I’ll give it a shot.

Just throwing another idea out there: use duplicated quotes `''` or `""` instead of escaping. This would avoid making ˋ\ˋ special.

> > It would need to be generalized to the more general case I just mentioned but yes I think it should be doable.
> > Sounds good? (thanks for the sample code)
> 
> Why do you care about arbitrary number of trailing backslashes? the last one can be special (ensure we do not confuse with an escaped trailing quote), the rest can be regular and not escaped.

This is related to the first point above: it is for proper identification of the end of a quoted string with a regex. If `'abc\\\'`resolves to `abc\\` (I think this is what you are suggesting) and `'abc\\'` resolves to `abc\'` then how do you get a quoted string with one backslash followed by a quote in the middle of the string? (desired output: `abc\'def`)

Reply:
> > This is not related to us using lexer modes there, right? it will just eliminate the need to use double backslash to get a backslash in quotes strings? Put up a PR and we'll see.
> 
> Correct. I’ll give it a shot.

Done in 2369a0638

It does make the un-escaping code (and its `with_quotes()` counterpart) more complex, but personally I think the benefits in terms of user-friendliness are worth it.

Reply:
I also added a few more tests in 3bf6156b6 to better test what happens with more complex interpolations inside quoted strings

Reply:
Aaaand a doc improvement to better explain how to pass interpolations through the command line (in case someone would actually want to do it): 00cd37f9a

Reply:
> This is related to the first point above: it is for proper identification of the end of a quoted string with a regex. If `'abc\\\'`resolves to `abc\\` (I think this is what you are suggesting) and `'abc\\'` resolves to `abc\'` then how do you get a quoted string with one backslash followed by a quote in the middle of the string? (desired output: `abc\'def`)

Regex is an implementation detail here.
With my implementation, this passes:
```
(r"abc\\'def", r"abc\'def"),
```
I will hold reviewing the changes until we clear this point up.

Reply:
> > This is related to the first point above: it is for proper identification of the end of a quoted string with a regex. If `'abc\\\'`resolves to `abc\\` (I think this is what you are suggesting) and `'abc\\'` resolves to `abc\'` then how do you get a quoted string with one backslash followed by a quote in the middle of the string? (desired output: `abc\'def`)
> 
> Regex is an implementation detail here.
> With my implementation, this passes:
> 
> ```
> (r"abc\\'def", r"abc\'def"),
> ```

Your implementation is operating at a lower level, *after* the content of the string has been extracted (in your example, the characters `abc\\'def`, before un-escaping).

Stepping back to the higher-level grammar parsing, this content must from a quoted string, that one might want to look like this (again, these are the characters inside the string):
```
'abc\\'def'
```
I consider that such a quoted string would actually be invalid, because `\\'` should mark the end of the string, since we also want the following quoted string to be valid:
```
'abc\\'  <== should resolve to: abc\
```

It *might* be possible to somehow detect the end of the quoted string so that both examples above work, but this seems tricky (maybe I'm not seeing it?). It kinda worked in the previous implementation but at the expense of bugs like #1600.

Reply:
> 
```
'abc\\'def'
```
> I consider that such a quoted string would actually be invalid, because \\' should mark the end of the string, since we also want the following quoted string to be valid:

If we treat `\'` as an escaped quote that cannot close a string,  and the preceding `\` as a regular backslash, this string is not invalid.

If the low level lexer can correctly tokenize all the inputs we are talking about, I am not sure what is the problem.
The logic you are applying is operating on the characters in the quoted string token, right?
Here is an example that is closer to that for single quoted string, including `'abc\\'  <== should resolve to: abc\` case

The only real change is that now forward is also stripping the outer quotes.
I renamed forward to unquote and backward to quote.
This logic does not handle negative cases, it assumes the lexer would fail to produce a token if the string is illegal.
Example of an illegal string I am expecting the lexer to fail on: `'a\'` (It's illegal because it's not closed).

```python
from pytest import mark


def unquote(s: str) -> str:
    assert s[0] == s[-1] == "'"
    s = s[1:-1]
    if s.endswith("\\\\"):
        s1 = s[0:-2].replace(r"\'", "'") + "\\"
    else:
        s1 = s.replace(r"\'", "'")
    return s1


def quote(s):
    if s.endswith("\\"):
        s1 = s.replace("'", r"\'") + "\\"
    else:
        s1 = s.replace("'", r"\'")
    return f"'{s1}'"


@mark.parametrize(
    ("input_", "expected"),
    [
        (r"'ab'", r"ab"),
        (r"'a\'b'", r"a'b"),
        (r"'ab\''", "ab'"),
        (r"'\ab'", r"\ab"),
        (r"'ab\\'", r"ab\ ".strip()),
        (r"'ab\\''", r"ab\'"),
        (r"'abc\\'def'", r"abc\'def"),
        (r"'abc\\'", r"abc\ ".strip()),
    ],
)
class TestEscape:
    def test_unquote(self, input_, expected):
        res = unquote(input_)
        assert res == expected

    def test_round_trip(self, input_, expected):
        res = unquote(input_)
        assert res == expected
        back = quote(res)
        assert back == input_
```

Reply:
> > 
> 
> ```
> 'abc\\'def'
> ```
> 
> > I consider that such a quoted string would actually be invalid, because \' should mark the end of the string, since we also want the following quoted string to be valid:
> 
> If we treat `\'` as an escaped quote that cannot close a string, and the preceding `\` as a regular backslash, this string is not invalid.
> 
> If the low level lexer can correctly tokenize all the inputs we are talking about, I am not sure what is the problem.

The problem *is* to get the lexer to correctly tokenize the inputs. What if you have the following override string (these are the characters in the string):
```
["abc\\", "def"]
```
=> how does it know that `"abc\\"` should be a single quoted value if we also allow `"abc\\", "` to be valid? I'm not saying it's impossible to achieve but I feel like it would make the lexer and/or parser significantly more complex, and may be heavier to maintain over time. I'd rather stay away from this unless you can suggest a satisfactory solution to this problem.

Reply:
I rebased on top of master and added 499802019 which contains the comments explaining the lexer regex for QUOTED_VALUE